Mergin/version_viewer_dialog.py

```diff
@@ -0,0 +1,639 @@
+        self.version = version
+
+    def run(self):
+        version_info = self.mc.project_version_info(self.mp.project_id(), version=f"v{self.version}")
+
+        files_updated = version_info["changes"]["updated"]
+
+        # if file not in project_version_info # skip as well
+        if not version_info["changesets"]:
+            self.finished.emit("This version does not contain changes in the project layers.")
+            return
+
+        files_updated = [f for f in files_updated if is_versioned_file(f["path"])]
+
+        if not files_updated:
+            self.finished.emit("This version does not contain changes in the project layers.")
+            return
+
+        has_history = any("diff" in f for f in files_updated)
+        if not has_history:
+            self.finished.emit("This version does not contain changes in the project layers.")
+            return
+
+        for f in files_updated:
+            if self.isInterruptionRequested():
+                return
+
+            if "diff" not in f:
+                continue
+            file_diffs = self.mc.download_file_diffs(self.mp.dir, f["path"], [f"v{self.version}"])
+            full_gpkg = self.mp.fpath_cache(f["path"], version=f"v{self.version}")
+            if not os.path.exists(full_gpkg):
+                self.mc.download_file(self.mp.dir, f["path"], full_gpkg, f"v{self.version}")
+
+        if self.isInterruptionRequested():
+            self.quit()
+            return
+
+        self.finished.emit("")
+
+
+class VersionsFetcher(QThread):
+
+    finished = pyqtSignal(list)
+
+    def __init__(self, mc: MerginClient, project_path, model: VersionsTableModel):
+        super(VersionsFetcher, self).__init__()
+        self.mc = mc
+        self.project_path = project_path
+        self.model = model
+
+        self.current_page = 1
+        self.per_page = 50
+
+        version_count = self.mc.project_versions_count(self.project_path)
+        self.nb_page = math.ceil(version_count / self.per_page)
+
+    def run(self):
+        self.fetch_another_page()
+
+    def has_more_page(self):
+        return self.current_page <= self.nb_page
+
+    def fetch_another_page(self):
+        if self.has_more_page() == False:
+            return
+        page_versions, _ = self.mc.paginated_project_versions(
+            self.project_path, self.current_page, per_page=self.per_page, descending=True
+        )
+        self.model.add_versions(page_versions)
+
+        self.current_page += 1
+
+
+class VersionViewerDialog(QDialog):
+    """
+    The class is constructed in a way that the flow of the code follow the flow the UI
+    The UI is read from left to right and each splitter is read from top to bottom
+
+    The __init__ method follow this pattern after varaible initiatlization
+    the methods of the class also follow this pattern
+    """
+
+    def __init__(self, mc, parent=None):
+
+        QDialog.__init__(self, parent)
+        self.ui = uic.loadUi(ui_file, self)
+
```

Comment:
`QgsGui.instance().enableAutoGeometryRestore(self)` should allow for QGIS to resize the dialog to it's last size/pos

Reply:
Added in latest commits as well